airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/global_substream_cursor.py

```diff
@@ -0,0 +1,277 @@
+        does not have parent streams, this step will be skipped due to the default PartitionRouter implementation.
+
+        Args:
+            stream_state (StreamState): The state of the streams to be set. The format of the stream state should be:
+                {
+                    "state": {
+                        "last_updated": "2023-05-27T00:00:00Z"
+                    },
+                    "parent_state": {
+                        "parent_stream_name": {
+                            "last_updated": "2023-05-27T00:00:00Z"
+                        }
+                    },
+                    "lookback_window": 132
+                }
+        """
+        if not stream_state:
+            return
+
+        if "lookback_window" in stream_state:
+            self._lookback_window = stream_state["lookback_window"]
+            self._inject_lookback_into_stream_cursor(stream_state["lookback_window"])
+
+        self._stream_cursor.set_initial_state(stream_state["state"])
+
+        # Set parent state for partition routers based on parent streams
+        self._partition_router.set_initial_state(stream_state)
+
+    def _inject_lookback_into_stream_cursor(self, lookback_window: int) -> None:
+        """
+        Modifies the stream cursor's lookback window based on the duration of the previous sync.
+        This adjustment ensures the cursor is set to the minimal lookback window necessary for
+        avoiding missing data.
+
+        Parameters:
+            lookback_window (int): The lookback duration in seconds to be set, derived from
+                                   the previous sync.
+
+        Raises:
+            ValueError: If the cursor does not support dynamic lookback window adjustments.
+        """
+        if hasattr(self._stream_cursor, "set_runtime_lookback_window"):
+            self._stream_cursor.set_runtime_lookback_window(lookback_window)
+        else:
+            raise ValueError("The cursor class for Global Substream Cursor does not have a set_runtime_lookback_window method")
+
+    def observe(self, stream_slice: StreamSlice, record: Record) -> None:
+        self._stream_cursor.observe(StreamSlice(partition={}, cursor_slice=stream_slice.cursor_slice), record)
+
+    def close_slice(self, stream_slice: StreamSlice, *args: Any) -> None:
+        """
+        Close the current stream slice.
+
+        This method is called when a stream slice is completed. For the global parent cursor, we close the child cursor
+        only after reading all slices. This ensures that we do not miss any child records from a later parent record
+        if the child cursor is earlier than a record from the first parent record.
+
+        Args:
+            stream_slice (StreamSlice): The stream slice to be closed.
+            *args (Any): Additional arguments.
+        """
+        if isinstance(stream_slice, GlobalCursorStreamSlice):
+            if stream_slice.last_slice:
```

Comment:
How will this translate to Concurrent CDK? It feels like it only works if the last generated slice is also the last one that is processed. I feel this might hinder the project `concurrent low-code` we have this cycle. Could we use another logic to understand if all the slices have been processed?

Reply:
I will rewrite it to use counter and lock for the `close_slice` method

Reply:
I don't think this is only a question of lock. I fear that as we process slice concurrently, this could happen:
1. yield first slice
2. yield last slice
3. last slice gets processed and closed
4. first slice gets processed and close

In this case, if the connectors fails between 3 and 4, the state saved would be wrong and it would lead to data loss

Reply:
EDIT: I understand what you mean by "counter" now. We will need a lock for that yes. Thanks!

Reply:
Fixed